onnxruntime/core/providers/cpu/reduction/reduction_ops.cc

```diff
@@ -455,6 +404,101 @@ void DropDimensions(const std::vector<int64_t>& input_shape, const std::vector<i
   }
 }
 
+FastReduceKind OptimizeShapeForFastReduce(const std::vector<int64_t>& input_shape,
+                                          const std::vector<int64_t>& reduced_axes,
```

Comment:
Some comments explaining the high level approach would be helpful to a developer who isn't familiar with the background of this change.

Reply:
The comment was in the header. I let it in the header. Should I copy it?

Reply:
In the header is fine now that the comment is attached to the declaration of OptimizeShapeForFastReduce.